sdk/core/src/context.rs

```diff
@@ -1,11 +1,177 @@
+use std::any::{Any, TypeId};
+use std::collections::HashMap;
+use std::sync::Arc;
+
 /// Pipeline execution context.
-#[derive(Clone, Debug, Default)]
+#[derive(Clone, Debug)]
 pub struct Context {
-    _priv: (),
+    pub type_map: HashMap<TypeId, Arc<dyn Any + Send + Sync>>,
+}
+
+impl Default for Context {
+    fn default() -> Self {
+        Self::new()
+    }
 }
 
 impl Context {
+    /// Creates a new, empty Context.
     pub fn new() -> Self {
-        Self::default()
+        Self {
+            type_map: HashMap::new(),
+        }
+    }
+
+    /// Inserts or replaces an entity in the type map. If an entity with the same type was displaced
+    /// by the insert, it will be returned to the caller.
+    pub fn insert_or_replace<E>(&mut self, entity: E) -> Option<Arc<E>>
+    where
+        E: Send + Sync + 'static,
+    {
+        self.type_map
+            .insert(TypeId::of::<E>(), Arc::new(entity))
+            .map(|displaced| displaced.downcast().unwrap())
+    }
+
+    /// Inserts an entity in the type map. If the an entity with the same type signature is
+    /// already present it will be silently dropped. This function returns a mutable reference to
+    /// the same Context so it can be chained to itself.
+    pub fn insert<E>(&mut self, entity: E) -> &mut Self
+    where
+        E: Send + Sync + 'static,
+    {
+        self.type_map.insert(TypeId::of::<E>(), Arc::new(entity));
+
+        self
+    }
+
+    /// Removes an entity from the type map. If found it will be returned.
+    pub fn remove<E>(&mut self) -> Option<Arc<E>>
+    where
+        E: Send + Sync + 'static,
+    {
+        self.type_map
+            .remove(&TypeId::of::<E>())
+            .map(|removed| removed.downcast().unwrap())
+    }
+
+    /// Returns a reference of the entity of the specified type signature, if exists. In there is
+    /// no entity with the specific type signature, `None` is returned instead.
+    pub fn get<E>(&self) -> Option<&E>
+    where
+        E: Send + Sync + 'static,
+    {
+        self.type_map
+            .get(&TypeId::of::<E>())
+            .map(|item| item.downcast_ref())
+            .flatten()
+    }
+
+    /// Returns the number of entities in the type map.
+    pub fn len(&self) -> usize {
+        self.type_map.len()
+    }
+
+    /// Iterates the type map contains.
+    pub fn iter(&self) -> std::collections::hash_map::Iter<'_, TypeId, Arc<dyn Any + Send + Sync>> {
```

Comment:
I think we should return `impl Iter` instead of the concrete type. Makes changing things in the future easier.

Reply:
I think I will remove the iterator code altogether. It's not really needed and exposes the underlying `Any` (which we should not do).

Reply:
If needed we can always add it in the future!